cmd/kubeapps-apis/core/packages/v1alpha1/packages_fan_in.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2021-2022 the Kubeapps contributors.
+// SPDX-License-Identifier: Apache-2.0
+package v1alpha1
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	packages "github.com/vmware-tanzu/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+	"github.com/vmware-tanzu/kubeapps/cmd/kubeapps-apis/plugins/pkg/paginate"
+)
+
+const CompleteToken = -1
+
+// summaryWithOffsets is the channel type for the results of the combined
+// core results after fanning in from the plugins.
+type summaryWithOffsets struct {
+	availablePackageSummary *packages.AvailablePackageSummary
+	categories              []string
+	nextItemOffsets         map[string]int
+	err                     error
+}
+
+// fanInAvailablePackageSummaries fans in the results from the separate plugins
+// to the return channel.
+//
+// Each plugin handles the request in a separate go-routine while this function
+// uses the fan-in pattern to merge those results, sending the next result back
+// down the return channel until the request is satisfied. Importantly, each
+// result is accompanied by the current next item offsets for each plugin so
+// that the caller can generate a next page token that is able to encode the
+// offsets for each plugin. The next request the begins each plugin where it
+// left off for the last.
+//
+// Plugins generally do not use snapshots of the actual data, so, similar to the
+// pagination of individual plugins, it will be possible that this returns
+// duplicates or missing data if data is added or removed between paginated
+// requests.
+func fanInAvailablePackageSummaries(ctx context.Context, pkgPlugins []pkgPluginWithServer, request *packages.GetAvailablePackageSummariesRequest) (<-chan summaryWithOffsets, error) {
+	summariesCh := make(chan summaryWithOffsets)
+
+	corePageSize := int(request.GetPaginationOptions().GetPageSize())
+	// We'll request a bit more than pageSize / n from each plugin. If the page
+	// size is 10 and we have 3 plugins, request 5 items from each to start.
+	pluginPageSize := corePageSize
+	if len(pkgPlugins) > 1 {
+		pluginPageSize = pluginPageSize / (len(pkgPlugins) - 1)
```

Comment:
What error propagation? Not sure which error you mean.

If `pluginPageSize` is an `int32` here (rather than `int`), then I need to explicitly cast the denominator as `int32`. Not sure what a ceiling function would achieve here? It's ints all the way down :P 

Looks like maybe you thought it was being preconverted *from a float* to an int to get integer division, rather than converting from `int32` to `int` to be able to use the types together? Not sure.

Reply:
Ahhhh, got it. Yep, I thought the division was returning a float instead of just performing the integer division. Thanks for pointing this out. Ignore the comment then :P 
PS: With error propagation I meant [sth like this](https://en.wikipedia.org/wiki/Propagation_of_uncertainty).

Reply:
Oh... that kind of error propagation (I'm so used to thinking about propagating errors in go code, in this PR through the channel, that my brain was looking for something along those lines that you might mean here, and I couldn't see an error :) )